Rename collection to item

koechkevin opened the pull request:
## Description
[MONGOOSE] Warning: `collection` is a reserved schema pathname and may break some functionality.
Fixes #485 

The following commands  were added
- Generate migrations file `pnpm migrate-mongo create {name_of_migrations}`.
- Migrate `turbo run db-migrate --filter=charterafrica`
- Rollback `turbo run db-rollback --filter=charterafrica`

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)


## Screenshots

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

kilemensi commented:
@koechkevin, when you say:
>We will need to update database column i.e
>

When/How are we supposed to be doing that? That's the crux of what you were supposed to figure out.

For example, did you look at the Payload migration [example](https://github.com/payloadcms/migrate-mongo-example)?

koechkevin commented:
> 



> @koechkevin, when you say:
> 
> > We will need to update database column i.e
> 
> When/How are we supposed to be doing that? That's the crux of what you were supposed to figure out.
> 
> For example, did you look at the Payload migration [example](https://github.com/payloadcms/migrate-mongo-example)?

@kilemensi I thought the scope of this task was only to have the name collection renamed and the existing data updated but I can update this PR as per the link you shared. I see it works the same way as _Sequelize_ normally handles migrations.

kilemensi commented:
> @kilemensi I thought the scope of this task was only to have the name collection renamed and the existing data updated but I can update this PR as per the link you shared. I see it works the same way as _Sequelize_ normally handles migrations.
>

You're right on the scope, it's the how to go about it that I was questioning @koechkevin i.e. having to manually update DEV and then PROD DBs. Isn't it better to automate it?

koechkevin commented:
@kilemensi 

I have updated this PR.
- Added `migrate-mongo` dependancy.
- Added script `db-migrate` for up migrations and `db-rollback` to undo migrations.

kilemensi reviewed (commented):
Do we need to loop and update one doc at a time @koechkevin? We can update many or something?

koechkevin commented:
> Do we need to loop and update one doc at a time @koechkevin? We can update many or something?

@kilemensi 
tried something like this 

```js
updateMany(
  {},
  { $rename: { "blocks.$[].items.$[].collection": "blocks.$[].items.$[].item" } }
)
```

and got this error
`The source field for $rename may not be dynamic: blocks.$[].items.$[].collection`

saminegash reviewed (commented):
I think for the renaming  rather renaming the collection to items I believe it is better to fist create Items collection using proper migration then copy everthing from the collection table to items table then update the usage of collection to items after that deleting the collection database is a good way to handle this migration

saminegash commented:
> I think for the renaming rather renaming the collection to items I believe it is better to fist create Items collection using proper migration then copy everthing from the collection table to items table then update the usage of collection to items after that deleting the collection database is a good way to handle this migration

If we do this I think we are going to have zero failed jobs on production

kilemensi commented:
> I think for the renaming rather ...
>
@koechkevin correct me if I'm wrong but we're not trying to rename a collection @saminegash. There is a **field** in a block or collection that is named `collection`. This PR is about renaming that field.

saminegash commented on apps/charterafrica/migrations/20230717085545-update_blocks_rename_collection_to_item.js:
```diff
@@ -0,0 +1,71 @@
+      .find()
+      .forEach((doc) => {
+        if (doc.blocks && doc.blocks.length > 0) {
+          doc.blocks.forEach((block) => {
```
here you have to use for of rather than foreach as foreach doesn't support blocking

koechkevin replied:
@saminegash not sure why you would propose a for...of over forEach when we do not need to break/continue out of loop.

koechkevin commented:
@kilemensi @saminegash I found a way to rename nested arrays using [Aggregation pipeline](https://www.mongodb.com/docs/manual/tutorial/update-documents-with-aggregation-pipeline). Tested and works fine

kilemensi commented on apps/charterafrica/turbo.json:
```diff
@@ -70,6 +70,8 @@
       ]
-    }
+    },
+    "db-rollback": {},
+    "db-migrate": {}
```
Do we need these in turbo right now?

kilemensi reviewed (approved):
🚀
